File: python/samples/gmod_operations.py
#!/usr/bin/env python3
"""GMOD Operations - Vista SDK Python.

This example demonstrates working with GMOD (Generic Product Model):
- Parsing and traversing GMOD paths
- Exploring node hierarchies
- Working with node metadata
- Version conversion operations
"""

from vista_sdk.gmod_node import GmodNode
from vista_sdk.gmod_path import GmodPath
from vista_sdk.traversal_handler_result import TraversalHandlerResult
from vista_sdk.vis import VIS
from vista_sdk.vis_version import VisVersion


def main() -> None:  # noqa : C901
    """Demonstrate GMOD operations."""
    print("=== GMOD Operations Example ===\n")

    # Initialize VIS
    vis = VIS()
    version = VisVersion.v3_4a
    gmod = vis.get_gmod(version)

    # 1. Basic GMOD path operations
    print("1. Basic GMOD Path Operations...")

    sample_paths = [
        "411.1",
        "411.1/C101.31-2",
        "612.21-1/C701.13/S93",
        "1021.1i-6P/H123",
    ]

    for path_str in sample_paths:
        try:
            path = gmod.parse_path(path_str)
            # or
            path = GmodPath.parse(path_str, arg=version)
            node = path.node
            print(f"   Path: {path_str}")
            print(f"     → Code: {node.code}")
            print(f"     → Name: {node.metadata.name}")
            print(f"     → Common Name: {node.metadata.common_name}")
            print(f"     → Depth: {path.length}")
            print()
        except Exception as e:
            print(f"   ✗ Failed to parse {path_str}: {e}")

    # 2. Exploring node hierarchies
    print("2. Exploring Node Hierarchies...")

    root_path = gmod.parse_path("411.1")
    root_node = root_path.node

    print(f"   Root node: {root_node.code} - {root_node.metadata.common_name}")
    print("   Child nodes:")

    # List immediate children
    children = list(root_node.children)[:5]  # Show first 5 children
    for child in children:
        print(f"     - {child.code}: {child.metadata.common_name}")

    # 3. Path traversal and analysis
    print("\n3. Path Traversal and Analysis...")

    deep_path = gmod.parse_path("411.1/C101.31-2")

    print(f"   Analyzing path: {deep_path}")
    print("   Full path traversal:")
    common_names = {c[0]: c[1] for c in deep_path.get_common_names()}
    for depth, node in deep_path.get_full_path():
        if depth == 0:
            continue  # Skip root
        # Common names must be looked up as they often are a combination between parent and child  # noqa: E501
        common_name = common_names.get(depth)
        if common_name is None:
            common_name = node.metadata.common_name
        indent = "     " * (depth - 1)
        print(f"   {indent}Depth {depth}: {node.code} - {common_name}")

    # 4. Node properties and metadata
    print("\n4. Node Properties and Metadata...")

    sample_nodes = ["411.1", "C101.31", "S206"]

    for node_code in sample_nodes:
        # Access through try_get pattern...
        success, lookup_node = gmod.try_get_node(node_code)
        try:
            # Or through direct lookup
            lookup_node = gmod[node_code]
            if lookup_node is None:
                raise ValueError(f"Node {node_code} not found")
            print(f"   Node: {node_code}")
            print(f"     → Name: {lookup_node.metadata.name}")
            print(f"     → Common Name: {lookup_node.metadata.common_name}")
            print(f"     → Is Mappable: {lookup_node.is_mappable}")
            print(f"     → Has Children: {len(lookup_node.children) > 0}")
            print()
        except Exception as e:
            print(f"   ✗ Node {node_code} not found: {e}")

    # 5. Path validation and error handling
    print("\n5. Path Validation and Error Handling...")

    invalid_paths = ["invalid.node", "411.1/invalid-child", "999.999/does-not-exist"]

    for invalid_path in invalid_paths:
        try:
            success, result = gmod.try_parse_path(invalid_path)
            if not success or result is None:
                print(f"   ✗ Path '{invalid_path}' is invalid (as expected)")
            else:
                print(f"   ⚠ Path '{invalid_path}' unexpectedly parsed as: {result}")
        except Exception as e:
            print(f"   ✗ Exception parsing '{invalid_path}': {e}")

    # 6. GMOD version conversion (if available)
    print("\n6. GMOD Version Conversion...")

    try:
        conversion_tests = [
            ("111.3/H402", VisVersion.v3_7a, VisVersion.v3_8a),
            ("846/G203", VisVersion.v3_7a, VisVersion.v3_8a),
        ]

        for old_path_str, source_version, target_version in conversion_tests:
            try:
                # Parse the old path
                old_path = GmodPath.parse(old_path_str, arg=source_version)

                # Convert using VIS convert_path method
                new_path = vis.convert_path(source_version, old_path, target_version)
                if new_path:
                    # No changes are very common
                    print(f"   ✓ Converted: {old_path_str} → {new_path}")
                else:
                    print(f"   ✗ Conversion returned None for {old_path_str}")
            except Exception as e:
                print(f"   ✗ Conversion failed for {old_path_str}: {e}")

    except Exception as e:
        print(f"   ⚠ Version conversion not available: {e}")

    print("\n7. GMOD Traversal...")
    node_count = 0

    def simple_handler(_: list[GmodNode], __: GmodNode) -> TraversalHandlerResult:
        nonlocal node_count
        node_count += 1
        return TraversalHandlerResult.CONTINUE

    completed = gmod.traverse(simple_handler)
    print(f"   Traversal completed: {completed}, Nodes visited: {node_count}")
    first_leaf: GmodNode | None = None

    def early_stop_handler(_: list[GmodNode], node: GmodNode) -> TraversalHandlerResult:
        nonlocal first_leaf
        if node.is_leaf_node:
            first_leaf = node
            return TraversalHandlerResult.STOP
        return TraversalHandlerResult.CONTINUE

    completed = gmod.traverse(early_stop_handler)
    print(f"   Traversal completed: {completed}, First leaf node: {first_leaf}")

    start_node = gmod["411"]
    child_cound = 0

    def from_node_handler(
        _: list[GmodNode],
        __: GmodNode,
    ) -> TraversalHandlerResult:
        nonlocal child_cound
        child_cound += 1
        return TraversalHandlerResult.CONTINUE

    completed = gmod.traverse(start_node, from_node_handler)

    print(f"   Traversal completed: {completed}, Nodes under '411': {child_cound}")

    print("\n=== GMOD operations completed! ===")


if __name__ == "__main__":
    main()

File: js/samples/GmodOperations.ts
/*
    GMOD Operations - Vista SDK JS/TS

    This example demonstrates working with GMOD (Generic Product Model):
    - Parsing and traversing GMOD paths
    - Exploring node hierarchies
    - Working with node metadata
    - Version conversion operations
*/

import {
    GmodNode,
    TraversalHandlerResult,
    VIS,
    VisVersion,
} from "dnv-vista-sdk";

async function main() {
    console.log("=== GMOD Operations Example ===\n");

    const version = VisVersion.v3_4a;
    const vis = VIS.instance;
    const { gmod, locations } = await vis.getVIS(version);

    // 1. Basic GMOD path operations
    console.log("1. Basic GMOD Path Operations...");

    const samplePaths = [
        "411.1",
        "411.1/C101.31-2",
        "612.21-1/C701.13/S93",
        "1021.1i-6P/H123",
    ];

    for (const pathStr of samplePaths) {
        const path = gmod.tryParsePath(pathStr, locations);
        if (path) {
            const node = path.node;
            console.log(`   Path: ${pathStr}`);
            console.log(`     → Code: ${node.code}`);
            console.log(`     → Name: ${node.metadata.name}`);
            console.log(
                `     → Common Name: ${node.metadata.commonName ?? ""}`,
            );
            console.log(`     → Depth: ${path.length}`);
            console.log();
        } else {
            console.log(`   ✗ Failed to parse ${pathStr}`);
        }
    }

    // 2. Exploring node hierarchies
    console.log("2. Exploring Node Hierarchies...");

    const rootPath = gmod.parsePath("411.1", locations);
    const rootNode = rootPath.node;

    console.log(
        `   Root node: ${rootNode.code} - ${rootNode.metadata.commonName ?? ""}`,
    );
    console.log("   Child nodes:");

    const children = rootNode.children.slice(0, 5);
    for (const child of children) {
        console.log(`     - ${child.code}: ${child.metadata.commonName ?? ""}`);
    }

    // 3. Path traversal and analysis
    console.log("\n3. Path Traversal and Analysis...");

    const deepPath = gmod.parsePath("411.1/C101.31-2", locations);

    console.log(`   Analyzing path: ${deepPath}`);
    console.log("   Full path traversal:");
    const commonNames = new Map(
        deepPath.getCommonNames().map((c) => [c.depth, c.name]),
    );
    const fullPath = deepPath.getFullPath();
    for (let depth = 0; depth < fullPath.length; depth++) {
        if (depth === 0) continue; // Skip root
        const node = fullPath[depth];
        const indent = " ".repeat((depth - 1) * 5);
        const commonName =
            commonNames.get(depth) ?? node.metadata.commonName ?? "";
        console.log(`   ${indent}Depth ${depth}: ${node.code} (${commonName})`);
    }

    // 4. Node properties and metadata
    console.log("\n4. Node Properties and Metadata...");

    const sampleNodes = ["411.1", "C101.31", "S206"];

    for (const nodeCode of sampleNodes) {
        const node = gmod.tryGetNode(nodeCode);
        if (node) {
            console.log(`   Node: ${nodeCode}`);
            console.log(`     → Name: ${node.metadata.name}`);
            console.log(
                `     → Common Name: ${node.metadata.commonName ?? ""}`,
            );
            console.log(`     → Is Mappable: ${node.isMappable}`);
            console.log(`     → Has Children: ${node.children.length > 0}`);
            console.log();
        } else {
            console.log(`   ✗ Node ${nodeCode} not found`);
        }
    }

    // 5. Path validation and error handling
    console.log("5. Path Validation and Error Handling...");

    const invalidPaths = [
        "invalid.node",
        "411.1/invalid-child",
        "999.999/does-not-exist",
    ];

    for (const invalidPath of invalidPaths) {
        const result = gmod.tryParsePath(invalidPath, locations);
        if (!result) {
            console.log(`   ✗ Path '${invalidPath}' is invalid (as expected)`);
        } else {
            console.log(
                `   ⚠ Path '${invalidPath}' unexpectedly parsed successfully`,
            );
        }
    }

    // 6. GMOD version conversion
    console.log("\n6. GMOD Version Conversion...");

    const conversionTests: [string, VisVersion, VisVersion][] = [
        ["111.3/H402", VisVersion.v3_7a, VisVersion.v3_8a],
        ["846/G203", VisVersion.v3_7a, VisVersion.v3_8a],
    ];

    for (const [oldPathStr, sourceVersion, targetVersion] of conversionTests) {
        const sourceGmod = await vis.getGmod(sourceVersion);
        const sourceLocations = await vis.getLocations(sourceVersion);

        const oldPath = sourceGmod.tryParsePath(oldPathStr, sourceLocations);
        if (!oldPath) {
            console.log(`   ✗ Failed to parse ${oldPathStr}`);
            continue;
        }

        const newPath = await vis.convertPath(
            sourceVersion,
            oldPath,
            targetVersion,
        );
        if (newPath) {
            console.log(
                `   ✓ Converted: ${oldPathStr} (${sourceVersion}) → ${newPath} (${targetVersion})`,
            );
        } else {
            console.log(`   ✗ Conversion returned undefined for ${oldPathStr}`);
        }
    }

    // 7. GMOD Traversal
    console.log("\n7. GMOD Traversal...");

    // Simple traversal - count all nodes
    let nodeCount = 0;
    const completed = gmod.traverse((parents: GmodNode[], node: GmodNode) => {
        nodeCount++;
        return TraversalHandlerResult.Continue;
    });
    console.log(
        `   Traversal completed: ${completed}, Nodes visited: ${nodeCount}`,
    );

    // Traversal with early stop - find first leaf node
    let firstLeaf: GmodNode | undefined;
    const completed2 = gmod.traverse((parents: GmodNode[], node: GmodNode) => {
        if (node.isLeafNode) {
            firstLeaf = node;
            return TraversalHandlerResult.Stop;
        }
        return TraversalHandlerResult.Continue;
    });
    console.log(
        `   Traversal completed: ${completed2}, First leaf node: ${firstLeaf}`,
    );

    // Traversal from specific node
    const startNode = gmod.getNode("411");
    let childCount = 0;
    const completed3 = gmod.traverse(
        (parents: GmodNode[], node: GmodNode) => {
            childCount++;
            return TraversalHandlerResult.Continue;
        },
        { rootNode: startNode },
    );

    console.log(
        `   Traversal completed: ${completed3}, Child nodes visited from '411': ${childCount}`,
    );

    console.log("\n=== GMOD operations completed! ===");
}

main().catch(console.error);
